### ISSUE TITLE
Issue in compiling DFTB+ from 21.2

### ISSUE THREAD

**Issue by gavinwinter** [dftbplus/dftbplus#1010]
Running the dftb+ command with any input script, I am getting the following error for my compiled 21.2 version. It also states that DFTB+ was compiled without support for tblite library in my output, but I manually added that from the pre-compiled 21.2 version. Any ideas on what the issue may be? Thanks in advance.

```
ERROR STOP 

Error termination. Backtrace:
#0  0x7fd84e3e22ed in ???
#1  0x7fd84e3e2ed5 in ???
#2  0x7fd84e3e3f86 in ???
#3  0x55f24dd940fc in __dftbp_common_globalenv_MOD_abortprogram
	at /home2/gridsan/gawinter/projects/dftb/dftbplus-21.2/src/dftbp/common/globalenv.F90:195
#4  0x55f24de93062 in readxtbham
	at /home2/gridsan/gawinter/projects/dftb/dftbplus-21.2/src/dftbp/dftbplus/parser.F90:1802
#5  0x55f24de96b9a in readhamiltonian
	at /home2/gridsan/gawinter/projects/dftb/dftbplus-21.2/src/dftbp/dftbplus/parser.F90:1225
#6  0x55f24de96b9a in __dftbp_dftbplus_parser_MOD_parsehsdtree
	at /home2/gridsan/gawinter/projects/dftb/dftbplus-21.2/src/dftbp/dftbplus/parser.F90:218
#7  0x55f24dd98c88 in __dftbp_dftbplus_hsdhelpers_MOD_parsehsdinput
	at /home2/gridsan/gawinter/projects/dftb/dftbplus-21.2/src/dftbp/dftbplus/hsdhelpers.F90:46
#8  0x55f24dd90f2c in dftbplus
	at /home2/gridsan/gawinter/projects/dftb/dftbplus-21.2/app/dftb+/dftbplus.F90:30
#9  0x55f24dd8fc8e in main
	at /home2/gridsan/gawinter/projects/dftb/dftbplus-21.2/app/dftb+/dftbplus.F90:11

```

**Comment by gavinwinter** [comment]
The reason I had to compile rather than using the pre-compiled version was because of [this issue](https://github.com/dftbplus/dftbplus/issues/940), which has been fixed (but not in the latest stable release of 21.2). The pre-compiled version does not have that bug fix, so xTB-MD does not work. I tried building from the latest version as pulled from github, but that has other changes as well that are not stable. Specifically, there are other shared object dependencies related to ELPA that cannot be found upon compiling (so that is a separate issue).

**Comment by gavinwinter** [comment]
So I realized I just wasn't specifying the WITH_TBLITE flag to be TRUE in my `config.cmake`. However, even after specifying that tblite has issues when building. The only reason I am posting this here is because I know tblite is made by some of the same developers as DFTB+. I read on some forums that this may be a gcc compiler issue, but after changing my gcc compiler to 11.2, I still get this error:
```
[ 35%] Building Fortran object external/tblite/origin/CMakeFiles/tblite-lib.dir/src/tblite/xtb/h0.f90.o
/home/gridsan/gawinter/projects/dftb/dftbplus-21.2/external/tblite/origin/src/tblite/xtb/h0.f90:404:0:

                nao = msao(bas%cgto(jsh, jzp)%ang)

Error: 'msao' not specified in enclosing 'parallel'
/home/gridsan/gawinter/projects/dftb/dftbplus-21.2/external/tblite/origin/src/tblite/xtb/h0.f90:373:0:

       do img = 1, list%nnl(iat)

Error: enclosing 'parallel'
/home/gridsan/gawinter/projects/dftb/dftbplus-21.2/external/tblite/origin/src/tblite/xtb/h0.f90:447:0:

          do iao = 1, msao(bas%cgto(ish, izp)%ang)

Error: 'msao' not specified in enclosing 'parallel'
/home/gridsan/gawinter/projects/dftb/dftbplus-21.2/external/tblite/origin/src/tblite/xtb/h0.f90:443:0:

       do ish = 1, bas%nsh_id(izp)

Error: enclosing 'parallel'
/home/gridsan/gawinter/projects/dftb/dftbplus-21.2/external/tblite/origin/src/tblite/xtb/h0.f90:218:0:

                nao = msao(bas%cgto(jsh, jzp)%ang)

Error: 'msao' not specified in enclosing 'parallel'
/home/gridsan/gawinter/projects/dftb/dftbplus-21.2/external/tblite/origin/src/tblite/xtb/h0.f90:197:0:

       do img = 1, list%nnl(iat)

Error: enclosing 'parallel'
/home/gridsan/gawinter/projects/dftb/dftbplus-21.2/external/tblite/origin/src/tblite/xtb/h0.f90:295:0:

             nao = msao(bas%cgto(jsh, izp)%ang)

Error: 'msao' not specified in enclosing 'parallel'
/home/gridsan/gawinter/projects/dftb/dftbplus-21.2/external/tblite/origin/src/tblite/xtb/h0.f90:282:0:

       do ish = 1, bas%nsh_id(izp)

Error: enclosing 'parallel'
external/tblite/origin/CMakeFiles/tblite-lib.dir/build.make:958: recipe for target 'external/tblite/origin/CMakeFiles/tblite-lib.dir/src/tblite/xtb/h0.f90.o' failed
make[2]: *** [external/tblite/origin/CMakeFiles/tblite-lib.dir/src/tblite/xtb/h0.f90.o] Error 1
make[2]: *** Waiting for unfinished jobs....
CMakeFiles/Makefile2:2982: recipe for target 'external/tblite/origin/CMakeFiles/tblite-lib.dir/all' failed
make[1]: *** [external/tblite/origin/CMakeFiles/tblite-lib.dir/all] Error 2
Makefile:145: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by awvwgk** [comment]
Which compiler is used here? Looks like GCC 7, which is unsupported for DFTB+.

**Comment by gavinwinter** [comment]
Yes. When I first tried that I was using gcc 7.4, but upon upgrading to gcc 11.2, I got this error when compiling:
```
[ 24%] Building Fortran object external/dftd4/origin/CMakeFiles/dftd4-lib.dir/src/dftd4/reference.f90.o
/usr/bin/ld: cannot find /lib64/libpthread.so.0
/usr/bin/ld: cannot find /usr/lib64/libpthread_nonshared.a
collect2: error: ld returned 1 exit status
external/multicharge/origin/app/CMakeFiles/multicharge-exe.dir/build.make:100: recipe for target 'external/multicharge/origin/app/multicharge' failed
make[2]: *** [external/multicharge/origin/app/multicharge] Error 1
CMakeFiles/Makefile2:2840: recipe for target 'external/multicharge/origin/app/CMakeFiles/multicharge-exe.dir/all' failed
make[1]: *** [external/multicharge/origin/app/CMakeFiles/multicharge-exe.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 24%] Building Fortran object external/s-dftd3/origin/CMakeFiles/s-dftd3-lib.dir/src/dftd3/model.f90.o
[ 24%] Building Fortran object external/dftd4/origin/CMakeFiles/dftd4-lib.dir/src/dftd4/model.f90.o
[ 24%] Building Fortran object external/dftd4/origin/CMakeFiles/dftd4-lib.dir/src/dftd4/damping/rational.f90.o
[ 25%] Building Fortran object external/s-dftd3/origin/CMakeFiles/s-dftd3-lib.dir/src/dftd3/output.f90.o
[ 25%] Building Fortran object external/s-dftd3/origin/CMakeFiles/s-dftd3-lib.dir/src/dftd3/disp.f90.o
[ 25%] Building Fortran object external/s-dftd3/origin/CMakeFiles/s-dftd3-lib.dir/src/dftd3.f90.o
[ 25%] Building Fortran object external/dftd4/origin/CMakeFiles/dftd4-lib.dir/src/dftd4/param.f90.o
[ 25%] Building Fortran object external/dftd4/origin/CMakeFiles/dftd4-lib.dir/src/dftd4/output.f90.o
[ 25%] Building Fortran object external/dftd4/origin/CMakeFiles/dftd4-lib.dir/src/dftd4/disp.f90.o
[ 26%] Linking Fortran static library libs-dftd3.a
[ 26%] Building Fortran object external/multicharge/origin/test/CMakeFiles/multicharge-tester.dir/main.f90.o
[ 26%] Built target s-dftd3-lib
[ 26%] Linking Fortran executable multicharge-tester
/usr/bin/ld: cannot find /lib64/libpthread.so.0
/usr/bin/ld: cannot find /usr/lib64/libpthread_nonshared.a
collect2: error: ld returned 1 exit status
external/multicharge/origin/test/CMakeFiles/multicharge-tester.dir/build.make:161: recipe for target 'external/multicharge/origin/test/multicharge-tester' fail                                                ed
make[2]: *** [external/multicharge/origin/test/multicharge-tester] Error 1
CMakeFiles/Makefile2:2868: recipe for target 'external/multicharge/origin/test/CMakeFiles/multicharge-tester.dir/all' failed
make[1]: *** [external/multicharge/origin/test/CMakeFiles/multicharge-tester.dir/all] Error 2
[ 26%] Building Fortran object external/dftd4/origin/CMakeFiles/dftd4-lib.dir/src/dftd4.f90.o
[ 26%] Linking Fortran static library libdftd4.a
[ 26%] Built target dftd4-lib
Makefile:145: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by awvwgk** [comment]
Looks like GCC 11 is not properly installed than. I presume you have purged the build directory and/or the `CMakeCache.txt` after the upgrade?

**Comment by gavinwinter** [comment]
Yes. I ensured that I deleted the `_build` directory and `CMakeCache.txt` after the upgrade. I think it is just not able to find `libpthread` (or at least that's one of the errors). Is there a way for cmake to point to it (similar to how this would be done command line `-D SCALAPACK_LIBRARY=$HOME/.conda/envs/cmake/lib`)?

**Comment by gavinwinter** [comment]
For reference, I'm using [the version 21.2 from the stable release page](https://dftbplus.org/download/dftb-stable/), and it is the "source"

**Comment by aradi** [comment]
No, in theory, CMake is supposed to find it automatically, when a dependency on OpenMP is detected. (Our targets should depend on OpenMP.)

Two things:
* Could you try to make a compilation without OpenMP (passing `-DWITH_OMP=0` to cmake). Just to test, whether it is really just an OpenMP related problem.
* If yes, you may try to call cmake with `FFLAGS="-fopenmp" cmake ...` to set explicitely OpenMP parallelisation via the according compiler flag.

**Comment by gavinwinter** [comment]
Ok. Got it. I am now getting this error with the `WITH_OMP` flag set to `FALSE` in my `config.cmake`.
```
(cmake) gawinter@login-2:~/projects/dftb/dftbplus-21.2$ FC=gfortran CC=gcc cmake -DCMAKE_INSTALL_PREFIX=$HOME/projects/dftb/dftb+ -B _build .
-- Reading global build config file: /home/gridsan/gawinter/projects/dftb/dftbplus-21.2/config.cmake
-- The Fortran compiler identification is GNU 11.2.0
-- The C compiler identification is GNU 11.2.0
-- Detecting Fortran compiler ABI info
-- Detecting Fortran compiler ABI info - done
-- Check for working Fortran compiler: /home/gridsan/gawinter/.conda/envs/cmake/bin/gfortran - skipped
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /home/gridsan/gawinter/.conda/envs/cmake/bin/gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Build type: RelWithDebInfo (default single-config)
-- Reading build environment specific toolchain file: /home/gridsan/gawinter/projects/dftb/dftbplus-21.2/sys/gnu.cmake
-- Flags for Fortran-compiler (build type: RelWithDebInfo):    -g -O2 -funroll-all-loops
-- Flags for C-compiler (build type: RelWithDebInfo):
-- Found Git: /usr/bin/git (found version "2.17.1")
-- Looking for Fortran sgemm
-- Looking for Fortran sgemm - not found
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Could NOT find BLAS (missing: BLAS_LIBRARIES)
CMake Error at /home/gridsan/gawinter/.conda/envs/cmake/share/cmake-3.22/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find CustomBlas (missing: BLAS_LIBRARY)
Call Stack (most recent call first):
  /home/gridsan/gawinter/.conda/envs/cmake/share/cmake-3.22/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  cmake/Modules/FindCustomBlas.cmake:109 (find_package_handle_standard_args)
  CMakeLists.txt:105 (find_package)


-- Configuring incomplete, errors occurred!
See also "/home/gridsan/gawinter/projects/dftb/dftbplus-21.2/_build/CMakeFiles/CMakeOutput.log".
See also "/home/gridsan/gawinter/projects/dftb/dftbplus-21.2/_build/CMakeFiles/CMakeError.log".
```

**Comment by gavinwinter** [comment]
I tried to check that conda has the things in the fourth column. 
```
| x86_64 /      | GNU Fortran/C 11.2   | OpenMPI 4.1 | OpenBlas 0.3.18, |     |
| Linux         |                      |             | ScaLAPACK 2.1    |     |
|               |                      |             | ELSI 2.8.2       |     |
```
I was able to locate `libopenblas` and verify that it is the right version, but I'm not sure if I need to specify ``CMAKE_PREFIX_PATH`` as an environment variable to get cmake to recognize it.
```
# packages in environment at /home/gridsan/gawinter/.conda/envs/cmake:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       2_gnu    conda-forge
binutils_impl_linux-64    2.36.1               h193b22a_2    conda-forge
bzip2                     1.0.8                h7b6447c_0
c-ares                    1.18.1               h7f8727e_0
ca-certificates           2022.3.29            h06a4308_1
cmake                     3.22.1               h1fce559_0
expat                     2.4.4                h295c915_0
gcc                       11.2.0               h702ea55_9    conda-forge
gcc_impl_linux-64         11.2.0              h82a94d6_16    conda-forge
gfortran                  11.2.0               h8811e0c_9    conda-forge
gfortran_impl_linux-64    11.2.0              h7a446d4_16    conda-forge
kernel-headers_linux-64   2.6.32              he073ed8_15    conda-forge
krb5                      1.19.2               hac12032_0
ld_impl_linux-64          2.36.1               hea4e1c9_2    conda-forge
libblas                   3.9.0           13_linux64_openblas    conda-forge
libcurl                   7.82.0               h0b77cf5_0
libedit                   3.1.20210910         h7f8727e_0
libev                     4.33                 h7f8727e_1
libgcc-devel_linux-64     11.2.0              h0952999_16    conda-forge
libgcc-ng                 11.2.0              h1d223b6_16    conda-forge
libgfortran-ng            11.2.0              h69a702a_16    conda-forge
libgfortran5              11.2.0              h5c6108e_16    conda-forge
libgomp                   11.2.0              h1d223b6_16    conda-forge
liblapack                 3.9.0           13_linux64_openblas    conda-forge
libnghttp2                1.46.0               hce63b2e_0
libopenblas               0.3.18          pthreads_h8fe5266_0    conda-forge
libsanitizer              11.2.0              he4da1e4_16    conda-forge
libssh2                   1.9.0                h1ba5d50_1
libstdcxx-ng              11.2.0              he4da1e4_16    conda-forge
libuv                     1.40.0               h7b6447c_0
lz4-c                     1.9.3                h295c915_1
mpi                       1.0                       mpich
mpich                     4.0.2              h846660c_100    conda-forge
ncurses                   6.3                  h7f8727e_2
openssl                   1.1.1n               h7f8727e_0
rhash                     1.4.1                h3c74f83_1
scalapack                 2.2.0                hd931219_1    conda-forge
sysroot_linux-64          2.12                he073ed8_15    conda-forge
xz                        5.2.5                h7b6447c_0
zlib                      1.2.12               h7f8727e_2
zstd                      1.5.2                ha4553b6_0
```

**Comment by aradi** [comment]
Actually, you should always adjust/extend `CMAKE_PREFIX_PATH` according to the desired libraries, as otherwise CMake often fails to find them, or even worse, finds an other version of it...

**Comment by gavinwinter** [comment]
Got it. I was able to successfully compile dftb+ with tblite (without MPI). I had to have my path to libraries `$HOME/.conda/envs/my_env/lib` specified explicitly with `CMAKE_PREFIX_PATH`. However, this alone did not fix the issue. I installed libblas with conda, which appends numbers to the `.so` libraries and causes cmake to not be able to find the library. So, I had to create a symbolic link using `ln -s <original_name> <new_name>`, where `libblas.so.3` was the original name and `libblas.so` was the new name. Hopefully, this helps future users.
